crates/valence_entity/src/hitbox.rs

```diff
@@ -1,294 +1,544 @@
+pub struct EntityHitboxSettings {
+    /// Controls if a plugin should add hitbox component on each created entity.
+    /// Otherwise you should add hitbox component by yourself in order to use
+    /// it.
+    pub add_hitbox_component: bool,
+}
+
+impl Default for EntityHitboxSettings {
+    fn default() -> Self {
+        Self {
+            add_hitbox_component: true,
         }
     }
+}
 
-    fn item_frame(pos: DVec3, rotation: i32) -> Aabb {
-        let mut center_pos = pos + 0.5;
-
-        match rotation {
-            0 => center_pos.y += 0.46875,
-            1 => center_pos.y -= 0.46875,
-            2 => center_pos.z += 0.46875,
-            3 => center_pos.z -= 0.46875,
-            4 => center_pos.x += 0.46875,
-            5 => center_pos.x -= 0.46875,
-            _ => center_pos.y -= 0.46875,
-        };
+impl Plugin for HitboxPlugin {
+    fn build(&self, app: &mut App) {
+        app.init_resource::<EntityHitboxSettings>()
+            .configure_set(HitboxUpdateSet.in_base_set(CoreSet::PreUpdate))
+            .add_systems(
+                (
+                    update_constant_hitbox,
+                    update_warden_hitbox,
+                    update_area_effect_cloud_hitbox,
+                    update_armor_stand_hitbox,
+                    update_passive_child_hitbox,
+                    update_zombie_hitbox,
+                    update_piglin_hitbox,
+                    update_zoglin_hitbox,
+                    update_player_hitbox,
+                    update_item_frame_hitbox,
+                    update_slime_hitbox,
+                    update_painting_hitbox,
+                    update_shulker_hitbox,
+                )
+                    .in_set(HitboxUpdateSet),
+            )
+            .configure_set(HitboxComponentAddSet.in_base_set(CoreSet::PostUpdate))
+            .add_system(add_hitbox_component.in_set(HitboxComponentAddSet))
+            .configure_set(
+                InWorldHitboxUpdateSet
+                    .in_base_set(CoreSet::PreUpdate)
+                    .after(HitboxUpdateSet),
+            )
+            .add_system(update_in_world_hitbox.in_set(InWorldHitboxUpdateSet));
+    }
+}
 
-        let bounds = DVec3::from(match rotation {
-            0 | 1 => [0.75, 0.0625, 0.75],
-            2 | 3 => [0.75, 0.75, 0.0625],
-            4 | 5 => [0.0625, 0.75, 0.75],
-            _ => [0.75, 0.0625, 0.75],
-        });
+/// Size of hitbox. The only way to manipulate it without losing it on the next
+/// tick is using a marker entity. Marker entity's hitbox is never updated.
+#[derive(Component, Debug, PartialEq)]
+pub struct Hitbox(pub Aabb);
 
-        Aabb {
-            min: center_pos - bounds / 2.0,
-            max: center_pos + bounds / 2.0,
+#[derive(Component, Debug)]
+/// Hitbox, aabb of which is calculated each tick using it's position and
+/// [`Hitbox`]. In order to change size of this hitbox you need to change
+/// [`Hitbox`]
+pub struct InWorldHitbox(Aabb);
```

Comment:
Why have separate components for a "size" hitbox and an "in world" hitbox? This seems a bit more complicated than it needs to be.

My suggestion is to modify your `update_*` systems to change the hitbox as the `Position` changes. So instead of `Added<Hitbox>`, you query for `Position` and use `Changed<Position>`.

I can see how this setup would let you freely change the hitbox size without changing its position, but I'm not sure we really need that. Apologies if my previous comments were misinterpreted.

Reply:
How we ensure that the hitbox size is the same as the entity's? It is possible but the only way i found is more complicated as if we had a size of hitbox

Reply:
What's wrong with
```rust
fn update_zombie_hitbox(
    mut query: Query<
        (&mut Hitbox, &Position, &zombie::Baby),
        Or<(Changed<Position>, Changed<zombie::Baby>)>,
    >,
) {
    for (mut hitbox, pos, baby) in query.iter_mut() {
        hitbox.0 = Aabb::from_bottom_size(pos.0, child_hitbox(baby.0, [0.6, 1.95, 0.6].into()));
    }
}
```
Then `Hitbox` takes the place of `InWorldHitbox` and you can remove `InWorldHitbox`.

Reply:
But if hitbox is changed out of all update hitbox functions?

Reply:
I don't understand?

Reply:
As i think in your system is the problem of too many needless calculations (like for constant systems and other systems which are depend on `EntityKind`)

We can get rid of a `InWorldHitbox` if we do this kind of system:
```rust
fn update_in_world_hitbox(
    mut hitbox_query: Query<
        (&mut Hitbox, &OldPosition, &Position),
        Or<Changed<Position>>,
    >,
) {
    for (mut hitbox, old_pos, pos) in hitbox_query.iter_mut() {
        hitbox.0 = hitbox.0 - old_pos.0 + pos.0;
    }
}
```
If we want to change a size of hitbox, we do that:
```rust
impl Hitbox {
    fn change_size(&mut self, new_size: Aabb, old_pos: DVec3) {
        self.0 = new_size + old_pos;
    }
}
```

And also we should update all our systems to calculate hitbox from its old_pos, so `centered` function should be changed to like this (and also some systems that don't use centered should add old_pos to something):
```rust
impl Hitbox {
    fn centered(&mut self, size: DVec3, old_pos: DVec3) {
        self.0 = Aabb::from_bottom_size(DVec3::ZERO, size) + old_pos;
    }
}
```

But this solution is also complicated (for some methods we need to use `OldPosition`)

Reply:
> As i think in your system is the problem of too many needless calculations (like for constant systems and other systems which are depend on EntityKind)

Hmm, this looks like a classic case of premature optimization. Honestly I would do the simple approach right now and change it later if you can show that it's an improvement, because right now I'm not yet convinced this is actually faster.

BTW, your other solution involving `OldPosition` might be vulnerable to accumulated floating point errors.

Reply:
I've been thinking and came to this. With my approach entity's custom hitboxes (for now only marker entity) will be easier to implement, because `InWorldHitbox` depends on `Position` and `Hitbox` when in yours the user will need to implement it's own system to change `Hitbox` to match it's new position.

Reply:
@dyc3 Do you have any opinions on this matter? 😄

Reply:
I think that the current solution here, where we have both `Hitbox` and `InWorldHitbox` types, is the easiest to understand, and allows for the most flexibility for users. I really don't like the `update_in_world_hitbox` system proposed in this thread, it feels like a very easy source of bugs. 

I agree with the premature optimization sentiment. I would feel much more comfortable optimizing this if we had some real numbers and real use cases to reference. Otherwise, the current impl seems perfectly fine to me.

Reply:
Alright, sounds good. Thanks for the input.

The last thing I'd consider changing is the name of the components. Perhaps `InWorldHitbox` should be renamed to `Hitbox` and `Hitbox` to `HitboxShape`, just to avoid confusion and emphasize that the "in world" hitbox is what users are going to be looking for most of the time. Other than that, LGTM.